24.11 ZHF Hackathon report (#8)
nix-community/ZurichZHF: 2024-11-report/README.md

# Zurich 24.11 ZHF hackathon

On the last weekend of November 2024, the Zürich Friends of Nix hosted yet another two-day hackathon around the [Zero Hydra Failures release preparations for NixOS 24.11](https://discourse.nixos.org/t/lets-have-a-great-24-11-release-cycle/51549) at [Eastern Switzerland University of Applied Sciences](https://www.ost.ch/) (OST) right next to Lake Zürich.

> We're proud to have hosted the fifth such event in a row. It's designed to encourage people to get to know each other, hack on everything Nix, and shape the future of open source together.

Reviewer: ```suggestion
These events are designed to encourage people to get to know each other, hack on everything Nix, and shape the future of open source together. We are proud to have hosted the fifth such event in a row.
```

![](./DSC05782.jpg)

# Background and preparations

Preparations started a month after the previous hackathon, by publishing [the report](https://discourse.nixos.org/t/2024-05-25-26-saturday-sunday-zhf-24-05-hackathon-ost-zurich-rapperswil-jona/45380/2), evaluating feedback, gathering ideas, and planning what needs to be done for the next event.

> In particular we focused enabling more people to contribute, welcoming new members to the community, better learning to know each other, and working on hard problems together. This primarily involved improving outreach and consistency of communication and reminders, in order to help get the right people into one room.

Reviewer: ```suggestion
In particular we focused on enabling more people to contribute, welcoming new members to the community, better getting to know each other, and working on hard problems together. This primarily involved improving outreach and consistency of communication and reminders, in order to help get the right people together.
```

# Event

The event took place in two large classrooms and the hallway connecting them. It spanned Saturday and Sunday from 10:00 to 18:00. This time the organiser team was suprised to arrive to half a room full of people who already started hacking around 9:00.

> We had more than 40 attendees on Saturday and around 30 on Sunday, arriving from Switzerland and all over Europe: France, Germany, the Netherlands, even as far as Island. This time, part of the crowd was a release manager, a member of the [steering committee](https://github.com/NixOS/org/blob/main/doc/governance.md), multiple Nixpkgs maintainers and major contributors, members of some [community teams](https://nixos.org/community/#governance-teams), representatives of various companies operating in the Nix ecosystem, a few former and recent [Summer of Nix](https://github.com/ngi-nix/summer-of-nix) participants, as well as a number of beginners and more experienced contributors.

Reviewer 2: ```suggestion
We had more than 40 attendees on Saturday and around 30 on Sunday, arriving from Switzerland and all over Europe: France, Germany, the Netherlands, even as far as Iceland. We were glad to see several key people in the crowd: a release manager, a member of the [steering committee](https://github.com/NixOS/org/blob/main/doc/governance.md), multiple Nixpkgs maintainers and major contributors, [governance team members](https://nixos.org/community/#governance-teams), representatives of various companies operating in the Nix ecosystem, a few former and recent [Summer of Nix](https://github.com/ngi-nix/summer-of-nix) participants, as well as a number of beginners and more experienced contributors.
```

> We were very happy to welcome an extraordinary number of people new to Nix, meet again our regulars such as @infinisil@ roberth @imincik @gefla @Nebucatnetzer @das-g, and especially appreciated having @wegank @tfc @SynQ @hsjobeki @erictapen who were around to share their expertise.

Reviewer 2: ```suggestion
We were very happy to welcome an extraordinary number of people new to Nix as well as our regulars such as @infinisil, @roberth, @imincik, @gefla, @Nebucatnetzer and @das-g. We especially appreciated having @wegank, @tfc, @SynQ, @hsjobeki, and @erictapen around to share their expertise.
```

On Saturday:
- Alex (@ners) welcomed attendees
- Valentin (@fricklerhandwerk) provided an account of the history of governance in the Nix ecosystem, including the conclusion of the constitutional assembly and recent formation of the steering committee, finally answering [one of his first questions on Discourse](https://discourse.nixos.org/t/how-is-the-nix-project-governed/5294)
- Silvan (@infinisil) gave his traditional and ever-updated introduction to squashing bugs for ZHF, noting a record-number of Nixpkgs maintainers at the event to help with reviews and merging
- Robert (@roberth) presented his highly promising work on [NixOps4](https://github.com/nixops4/nixops4), now famously known as -- technically correct -- "a tool to do things"

Well-caffeinated, attendees clustered into groups working on or discussing various issues, summarised below.

![](./DSC05633.jpg)

> Saturday's hacking ended with a group photo while discussions continued late into the evening over this edition's culinary feat: cheese fondue and Glühwein. On Sunday the hacking continued, with brave Nix evangelists who brought their friends or colleages receiving their honorary T-shirts ("I'm a special snowflake" and "🪟🍎🐧❄ One of these things is not like the others" went out of stock quickly).

Reviewer 2: ```suggestion
Saturday's hacking ended with a group photo while discussions continued late into the evening over this edition's culinary feat: cheese fondue and Glühwein. On Sunday the hacking continued, with brave Nix evangelists who brought their friends or colleagues receiving their honorary T-shirts ("I'm a special snowflake" and "🪟🍎🐧❄ One of these things is not like the others" went out of stock quickly).
```

Over the weekend, the group [created 28](https://github.com/search?q=org%3ANixOS+ZurichZHF+created%3A2024-11-23..2024-11-25&type=pullrequests) and [closed 24](https://github.com/search?q=org%3ANixOS+ZurichZHF+closed%3A2024-11-23..2024-11-25&type=pullrequests) issues or pull requests, an all time high! At the time of writing, the event series resulted in [108 created](https://github.com/search?q=org%3ANixOS+ZurichZHF+created%3A%3C2025-01-19&type=pullrequests) and [105 concluded](https://github.com/search?q=org%3ANixOS+ZurichZHF+closed%3A%3C2025-01-19&type=pullrequests) contributions.

Both by the numbers and by the vibes, this was the greatest Nix event in Zürich so far.

![](./DSC05825.jpg)

# Discussions

> - @tfc @SynQ @roberth @fricklerhandwerk had a fruitful exchange on increasing organisational maturity around the Nix community, and which benefits it would bring if we invested effort in striking a healthy balance between:

Reviewer 2: ```suggestion
- @tfc @SynQ @roberth @fricklerhandwerk had a fruitful exchange on increasing organisational maturity around the Nix community and what we would gain by striking a healthy balance between:
```

- @das-g looked into what would be needed to bump `python3Packages.pytest-rerunfailures` to its latest version (not for ZHF / NixOS 24.11, instead just to have it up-to-date in the rolling "`unstable`" channels), but filed the [PR](https://github.com/NixOS/nixpkgs/pull/358855) for that only after the event.
- @das-g helped / advised two or so first-time contributors with backporting fixes (= extracting the relevant changes as a patch), that upstream had published on their main branches but not released in a tagged version, yet. These fixes were needed to unbreak package builds for ZHF.
- @matthewcroughan continued hacking on his SCION setup

### Deployment systems exchange

![](./IMG_0314.jpg)

@hsjobeki (Clan) @elikoga (Thymis), @roberth @SynQ @fricklerhandwerk (Fediversity): Each of us works on a framework that has NixOS deployments at its core and exposes a user interface for letting people orchestrate that. Every project has a slightly different user group and goal they ultimately want to achieve. The underlying technical challenges the we are facing have a big overlap.

> We presented each of our architectures and how we aim to solve shared problems, to learn from each other and gain insights in fields that one of us have explored already. 

Reviewer 2: ```suggestion
We presented each of our architectures and how we aim to solve shared problems, to learn from each other, and to gain insights in fields that some of us have explored already. 
```

![](./IMG_0320.jpg)

### On user interfaces for NixOS configurations 

@hsjobeki and @infinisil worked out how NixOS-modules (options) could be rendered as a user-friendly UI. 

Problems downstream UI projects share:

- NixOS is a hard-to-work-with moving target, with options beeing renamed removed, changing behavior. Any abstraction on top of NixOS will likely fall apart if not maintained by the same people or continously kept in sync with downstream efforts.
- NixOS options cannot be set efficiently through a programmatically generated UI if they are not a subset of JSON.

> - Generating a UI from NixOS options ends up looking like a JSON schema editor. But not like the intuitive UI that non-techincal people would appreciate.

Reviewer 2: ```suggestion
- Generating a UI from NixOS options ends up looking like a JSON schema editor, not like an intuitive UI that non-technical people would appreciate.
```

> To brigde that gap a possible solutions could include options-metadata or type 'annotations' in Nix. We worked on option metadata, type merging, and improving `lib.types` as necessary preparation for that idea.

Reviewer 2: ```suggestion
To bridge that gap possible solutions could include options-metadata or type 'annotations' in Nix. We worked on option metadata, type merging, and improving `lib.types` as necessary preparation for that idea.
```

# Acknowledgements

This event was made possible by Prof. Dr. Farhad Mehta, professor of informatics at the OST, by accommodating us in a large, modern computer room with a beautiful lake view.

This event wouldn't be possible without the amazing event team around @ners and @john-rodewald. Great many thanks to
- @das-g for tending to the venue
- @matthewcroughan @maralorn @twesterhout @vkleen for preparing and schlepping supplies early in the morning

[Tweag](https://tweag.io), the Modus Create Open Source Program Office lead by Mathieu Boespflug (@mboes), and [Antithesis](https://antithesis.com), thanks to their ongoing sponsoring, enabled @infinisil participate as part of his work on the Nix ecosystem. The attendence of @roberth @SynQ and @fricklerhandwerk was supported by the European Commission's [NGI initiative](https://ngi.eu/) and [NLnet](https://nlnet.nl) through the [Fediversity project](https://fediversity.eu/).

Thanks to all the participants for your attendance and active involvement, and for providing [detailed feedback on the event](https://github.com/nix-community/developer-workshop/issues/7) so we can make it even better next time.

> And last but not least, many thanks to release NixOS 24.11 release managers @RossComputerGuy and @wegank for another great edition of our favorite Linux distribution!

Reviewer 2: ```suggestion
And last but not least, many thanks to NixOS 24.11 release managers @RossComputerGuy and @wegank for another great edition of our favorite Linux distribution!
```

# Looking ahead

The next event is already in the making, save the date:

> Saturday 24rd and Sunday 25th of May 2025

> Next time we want to make it even easier for people to work together on things they're interested in, and welcoming many more new members to the community. We hope to meet all of you again!

Reviewer 2: ```suggestion
Next time we want to make it even easier for people to work together on things they're interested in and to welcome many more new members to the community. We hope to meet all of you again!
```
